### DISCUSSION TITLE
Discussion on custom TCP plugins

### DISCUSSION THREAD

**Discussion by xiupengrong** [Kong/kong#8439] [Help] [answered]
Hi，all：
I want to proxy an RPC interface based on TCP coordination.
and implement the authentication function in the preread phase in Kong gateway, I don't know if this idea can be implemented, please give me some relevant reference or some direction to guide me, currently I get a data of userdata data type through ngx.req.socket() API, but I don't know how to use lua to parse this data type, I Is the current direction the right one?

**Comment by xiupengrong** [comment]
Is TCP capable of HTTP-like authentication?

**Comment by kikito** [comment, marked as answer]
Hello, the docs for `ngx.req.socket()` can be found here: https://github.com/openresty/lua-nginx-module#ngxreqsocket

The "userdata" that that returns you is a socket object. You can use it to send/receive data with the [receive](https://github.com/openresty/lua-nginx-module#tcpsockreceive), [receiveany](https://github.com/openresty/lua-nginx-module#tcpsockreceiveany) and [receiveuntil](https://github.com/openresty/lua-nginx-module#tcpsockreceiveuntil) methods.

```
function CustomHandler:preread(config)
  local socket, err = ngx.req.socket()
  if not socket then
    kong.log.error("Could not open socket: " .. err)
    return
  end
  local line, err, partial = sock:receive()
  if not line then
     kong.log.error("failed to read a line: ", err)
     return
 end
 kong.log("successfully read a line: ", line)
end
```

> Is TCP capable of HTTP-like authentication?

It is capable of authentication. I would not call it "HTTP-like" though.

What you can do is consume the initial bytes sent through the socket, check their contents, and decide to terminate the connection if they don't satisfy a criteria. Example:

```
local SECRET_PASSWORD="secret"
function CustomHandler:preread(config)
  ... (same code as above)
  if line ~= SECRET_PASSWORD then
    return kong.response.exit(400, "authentication failed")
  end
end
```

**Reply by xiupengrong** [reply]
Thank you very much for your reply, I will study ngx req socket related content carefully, thanks

**Reply by xiupengrong** [reply]
hi，kikito：
I have implemented tcp based authentication after your very good guidance, but I also want to implement tps or tpd flow limiting function, how to implement it, because I found that the plugin in preread stage will only run once in the whole tcp connection, and tcp is a long connection, so how to implement the flow limiting function, please give me some direction guidance, thank you very much! Is it possible to implement the plugin in the access phase of the tcp protocol?
